### PR TITLE
improvement: Add IoT credentials and data endpoint validation.

### PR DESCRIPTION
**Pull request by nikkhilmuthye**
**Issue #, if available:**

**Description of changes:**
Currently, the validation will happen on the startup of the Nucleus along with during deployments. The validation will check that the IoT credential and data endpoint has the same region as specified in the awsRegion configuration.

**Why is this change necessary:**

**How was this change tested:**
Added unit tests.

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [X] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java [hunk 1]
```diff
@@ -557,10 +562,38 @@ private void validateDeviceConfiguration(String thingName, String certificateFil
...
+     *
+     * @param awsRegion       the provided AWS region.
+     * @param iotCredEndpoint the provided IoT credentials endpoint
+     * @param iotDataEndpoint the providedIoT data endpoint
+     * @throws DeviceConfigurationException if the region is not valid or if the IoT endpoints do not have the AWS
+     *                                      region as a part of its URL.
+     */
+    public void validateEndpoints(String awsRegion, String iotCredEndpoint, String iotDataEndpoint)
+            throws DeviceConfigurationException {
+        // If the region value is empty/null, then try to get the region from the SDK lookup path
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java:577]: This comment doesn't seem correct, you're not doing any region lookup.

File: src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java [hunk 2]
```diff
@@ -557,10 +562,38 @@ private void validateDeviceConfiguration(String thingName, String certificateFil
...
+        if (Utils.isNotEmpty(awsRegion) && !Region.regions().contains(Region.of(awsRegion))) {
+            logger.atWarn().log("Error looking up AWS region");
+            throw new DeviceConfigurationException(String.format("Error looking up AWS region %s", awsRegion));
+        }
+        if (Utils.isNotEmpty(iotCredEndpoint) && !iotCredEndpoint.contains(awsRegion)) {
+            throw new DeviceConfigurationException(String.format("Error setting up IoT credentials endpoint %s",
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java:583]: this is far too generic. Tell the user what the problem is.

"IoT credential endpoint region does not match the AWS region of the device."

File: src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java [hunk 3]
```diff
@@ -33,17 +36,21 @@
...
 
     @Override
     @SuppressWarnings("PMD.PrematureDeclaration")
     public void activate(Map<String, Object> newConfig, Deployment deployment,
                          CompletableFuture<DeploymentResult> totallyCompleteFuture) {
         Map<String, Object> serviceConfig;
+        Map<String, Object> kernelConfig = null;
         if (newConfig.containsKey(SERVICES_NAMESPACE_TOPIC)) {
             serviceConfig = (Map<String, Object>) newConfig.get(SERVICES_NAMESPACE_TOPIC);
+            if (serviceConfig.containsKey(DEFAULT_NUCLEUS_COMPONENT_NAME)) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/activator/DefaultActivator.java:51]: why use the default name? Get the real name.

File: src/main/java/com/aws/greengrass/deployment/activator/KernelUpdateActivator.java [hunk 4]
```diff
@@ -61,6 +67,23 @@ public void activate(Map<String, Object> newConfig, Deployment deployment,
             return;
         }
 
+        if (newConfig.containsKey(SERVICES_NAMESPACE_TOPIC)) {
+            Map<String, Object> serviceConfig = (Map<String, Object>) newConfig.get(SERVICES_NAMESPACE_TOPIC);
+            if (serviceConfig.containsKey(DEFAULT_NUCLEUS_COMPONENT_NAME)) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/activator/KernelUpdateActivator.java:72]: same here. Get the proper name for the component.

File: src/main/java/com/aws/greengrass/deployment/activator/KernelUpdateActivator.java [hunk 5]
```diff
@@ -61,6 +67,23 @@ public void activate(Map<String, Object> newConfig, Deployment deployment,
...
+                String iotCredEndpoint = tryGetIoTCredEndpointFromNewConfig(kernelConfig);
+                String iotDataEndpoint = tryGetIoTDataEndpointFromNewConfig(kernelConfig);
+                try {
+                    deviceConfiguration.validateEndpoints(awsRegion, iotCredEndpoint, iotDataEndpoint);
+                } catch (DeviceConfigurationException e) {
+                    logger.atError().cause(e).log("Error validating IoT endpoints");
+                    return;
+                }
+            }
+        }
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/deployment/activator/KernelUpdateActivator.java]: Why do we need this code duplicated in both activator subclasses? we could extract it out to a common method in `DeploymentActivator.java`. Actually, I'd prefer the code to not live in the activators at all, but organize it alongside the rest of config validation here like this - https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/main/src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java#L122 
Also, can we complete the deployment future with `ComponentConfigurationValidationException` for consistency with existing validation? It helps convey the message more effectively to the customer

**Comment by shaguptashaikh** [comment]
Have you tried running any E2E tests with this change? I think the test that was using misconfiguration as a way to create a test scenario for cancellation will not work anymore -`GIVEN_deployment_in_dependency_resolution_stage_WHEN_cancel_event_received_THEN_deployment_should_be_cancelled`

File: src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java [hunk 6]
```diff
@@ -111,8 +117,12 @@ private void updateActionForDeployment(Map<String, Object> newConfig, Deployment
             return;
         }
         Map<String, Object> serviceConfig;
+        Map<String, Object> kernelConfig = null;
         if (newConfig.containsKey(SERVICES_NAMESPACE_TOPIC)) {
             serviceConfig = (Map<String, Object>) newConfig.get(SERVICES_NAMESPACE_TOPIC);
+            if (serviceConfig.containsKey(deviceConfiguration.getNucleusComponentName())) {
+                kernelConfig = (Map<String, Object>) serviceConfig.get(DEFAULT_NUCLEUS_COMPONENT_NAME);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java:124]: get the correct name?

File: src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java [hunk 7]
```diff
@@ -548,19 +553,50 @@ private void validateDeviceConfiguration(String thingName, String certificateFil
...
+     * @param awsRegion       the provided AWS region.
+     * @param iotCredEndpoint the provided IoT credentials endpoint
+     * @param iotDataEndpoint the providedIoT data endpoint
+     * @throws ComponentConfigurationValidationException if the region is not valid or if the IoT endpoints do not have
+     *                                                   the AWS region as a part of its URL.
+     */
+    public void validateEndpoints(String awsRegion, String iotCredEndpoint, String iotDataEndpoint)
+            throws ComponentConfigurationValidationException {
+        if (Utils.isNotEmpty(awsRegion) && !Region.regions().contains(Region.of(awsRegion))) {
+            logger.atWarn().log("Error looking up AWS region");
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java:579]: log the region

File: src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java [hunk 8]
```diff
@@ -548,19 +553,50 @@ private void validateDeviceConfiguration(String thingName, String certificateFil
...
+            throw new ComponentConfigurationValidationException(String.format("Error looking up AWS region %s",
+                    awsRegion));
+        }
+        if (Utils.isNotEmpty(iotCredEndpoint) && !iotCredEndpoint.contains(awsRegion)) {
+            throw new ComponentConfigurationValidationException(
+                    String.format("IoT credential endpoint region(%s) does not match the AWS region(%s) of the device",
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/deployment/DeviceConfiguration.java:585]: remove parenthesis and put a space after "region" (in both cases)

File: src/test/java/com/aws/greengrass/deployment/activator/KernelUpdateActivatorTest.java [hunk 9]
```diff
@@ -95,6 +95,8 @@ void GIVEN_deployment_activate_WHEN_takeConfigSnapshot_fails_THEN_deployment_fai
                 DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, mockIOE)));
     }
 
+
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/deployment/activator/KernelUpdateActivatorTest.java:98]: remove

File: src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java [hunk 10]
```diff
@@ -123,6 +132,20 @@ private void updateActionForDeployment(Map<String, Object> newConfig, Deployment
...
+            try {
+                deviceConfiguration.validateEndpoints(awsRegion, iotCredEndpoint, iotDataEndpoint);
+            } catch (ComponentConfigurationValidationException e) {
+                logger.atError().cause(e).log("Error validating IoT endpoints");
+                totallyCompleteFuture
+                        .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
+                return;
+            }
+        }
+
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java:148]: Nit - All this new code could be in a separate method, also kernel -> nucleus

**Review by shaguptashaikh — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]

**Review by shaguptashaikh — APPROVED** [review]

**Pull request merged**